pkg/operator/ceph/cluster/rbd/controller.go

```diff
@@ -0,0 +1,268 @@
+	// Fetch the cephRBDMirror instance
+	cephRBDMirror := &cephv1.CephRBDMirror{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, cephRBDMirror)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			logger.Debug("cephRBDMirror resource not found. Ignoring since object must be deleted.")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, errors.Wrap(err, "failed to get cephRBDMirror")
+	}
+
+	// The CR was just created, initializing status fields
+	if cephRBDMirror.Status == nil {
+		cephRBDMirror.Status = &cephv1.Status{}
+		cephRBDMirror.Status.Phase = k8sutil.Created
+		err := opcontroller.UpdateStatus(r.client, cephRBDMirror)
+		if err != nil {
+			return reconcile.Result{}, errors.Wrap(err, "failed to set status")
+		}
+	}
+
+	// Make sure a CephCluster is present otherwise do nothing
+	cephClusterSpec, isReadyToReconcile, cephClusterExists, reconcileResponse := opcontroller.IsReadyToReconcile(r.client, r.context, request.NamespacedName, controllerName)
+	if !isReadyToReconcile {
+		// This handles the case where the Ceph Cluster is gone and we want to delete that CR
+		//
+		// Also, only remove the finalizer if the CephCluster is gone
+		// If not, we should wait for it to be ready
+		// This handles the case where the operator is not ready to accept Ceph command but the cluster exists
+		if !cephRBDMirror.GetDeletionTimestamp().IsZero() && !cephClusterExists {
+			// Remove finalizer
+			err := opcontroller.RemoveFinalizer(r.client, cephRBDMirror)
+			if err != nil {
+				return reconcile.Result{}, errors.Wrap(err, "failed to remove finalizer")
+			}
+
+			// Return and do not requeue. Successful deletion.
+			return reconcile.Result{}, nil
+		}
+
+		logger.Debugf("CephCluster resource not ready in namespace %q, retrying in %q.", request.NamespacedName.Namespace, opcontroller.WaitForRequeueIfCephClusterNotReadyAfter.String())
+		return reconcileResponse, nil
+	}
+	r.cephClusterSpec = &cephClusterSpec
+
+	// Populate clusterInfo
+	// Always populate it during each reconcile
+	clusterInfo, _, _, err := mon.LoadClusterInfo(r.context, request.NamespacedName.Namespace)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to populate cluster info")
+	}
+	r.clusterInfo = clusterInfo
+
+	// Populate CephVersion
+	daemon := string(opconfig.MonType)
+	currentCephVersion, err := cephclient.LeastUptodateDaemonVersion(r.context, r.clusterInfo.Name, daemon)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrapf(err, "failed to retrieve current ceph %q version", daemon)
+	}
+	r.clusterInfo.CephVersion = currentCephVersion
+
+	// Set a finalizer so we can do cleanup before the object goes away
+	err = opcontroller.AddFinalizerIfNotPresent(r.client, cephRBDMirror)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to add finalizer")
+	}
+
+	// DELETE: the CR was deleted
+	if !cephRBDMirror.GetDeletionTimestamp().IsZero() {
```

Comment:
If there is no ceph cleanup, we don't have a reason for a finalizer. Can we not add a finalizer?

Reply:
Right.